presto-hive/src/main/java/io/prestosql/plugin/hive/HiveLocationService.java

```diff
@@ -64,7 +64,7 @@ public LocationHandle forNewTable(SemiTransactionalHiveMetastore metastore, Conn
         }
 
         // TODO detect when existing table's location is a on a different file system than the temporary directory
-        if (shouldUseTemporaryDirectory(session, context, targetPath, externalLocation)) {
+        if (withData && shouldUseTemporaryDirectory(session, context, targetPath, externalLocation)) {
```

Comment:
IMO this is solving problem at the wrong level of abstraction.
in particular i wonder why we can't solve it within HiveMetadata, not touching (and so: not introducing more comlexity) into lower-level parts of the hive connector.

Reply:
in fact, i wonder whether we could do the change on the engine side of Presto, so that connectors do not have to deal with this complexity. What prevents as from calling `createTable` instead of `beginCreateTable`, for `CTAS WITH NO DATA`

cc @martint

Reply:
Yes, it should be executed as a `createTable`. Semantically, a CREATE TABLE t AS (<subquery>) WITH [NO] DATA is equivalent to:

```
CREATE TABLE t (...)
```

with the schema derived from the subquery, followed by:

```
INSERT INTO t <subquery>
```

Reply:
> What prevents as from calling createTable instead of beginCreateTable, for CTAS WITH NO DATA

Are we sure we don't do that? This PR is about fixing `createTable`.

Reply:
Thanks @sopel39 for the clarification. Then i missing something as I do not understand why location services needed a change.

Reply:
It's not. It gets planned as a `CREATE TABLE AS (<subquery> LIMIT 0)`:

```java
        if (!create.isCreateTableAsSelectWithData()) {
            PlanNode root = new LimitNode(idAllocator.getNextId(), plan.getRoot(), 0L, false);
            plan = new RelationPlan(root, plan.getScope(), plan.getFieldMappings(), Optional.empty());
        }
```

Reply:
That's because anything that's "query-like" needs to go through the normal query execution path. All the non-sql actions are implemented as `DataDefinitionTask`s (long-standing issue that eventually we need to clean up -- everything should be planned as a "query", with some queries just using operators that interact purely with metadata APIs, others read data, etc.)

Reply:
Do we have an issue for that?
I think the overarching goal is nice to have, but the case of CTAS WITH NO DATA is one of the painpoints, because it's not the first time connector needs to special case it.

But OK, this answers the question whether engine can do it -- yes it can, but seemingly it's a complex thing.

So the remaining would be -- can HiveMetadata handle this, without affecting location service?

Reply:
Actually I found this when testing CREATE TABLE, not CTAS. The problem is in HiveMetadata.createTable(), the temp staging directory is created and set WriteMode to STAGE_AND_MOVE_TO_TARGET_DIRECTORY, but at commit time, there is no declareIntentionToWrite, so the staging directory is not deleted. If we want to let HiveMetadata handle this, I want to confirm in this no write case, if the locationHandle/targetPath should be still generated? If it is still needed, then we need to set WriteMode correctly anyway, so location service need a flag to differentiate these two case.

Reply:
> Actually I found this when testing CREATE TABLE, not CTAS. The problem is in HiveMetadata.createTable(), the temp staging directory is created and set WriteMode to STAGE_AND_MOVE_TO_TARGET_DIRECTORY, but at commit time, there is no declareIntentionToWrite, so the staging directory is not deleted.

I totally missed that! thanks for adding more color here, as i already started to think what's the actual problem we're solving here.
I recall people complaining on slack about left over temp files, but back then we did not realize this is so reproducible.
Would it be possible to test this?

Reply:
I can add a testcase, basically create a table, and then verify the temp staging directory is not created. Can you suggest which existing test file to add this?

Reply:
Hive Smoke test?

Reply:
@findepi  I realized the way current temp staging directory is created is with a randomUUID: 
        Path temporaryPath = new Path(temporaryRoot, randomUUID().toString());
For example, in my debugger, the temporaryPath looks like "/tmp/presto-ychen6/cddaa90a-f6ac-49b8-9813-3f38bd9e1978". So it will be impossible to find/verify this path does not exist after CREATE TABLE statement is done. I could not think of any good way to test this, can we merge it as-is?

Reply:
The staging location is configurable (`io.prestosql.plugin.hive.HiveSessionProperties#TEMPORARY_STAGING_DIRECTORY_ENABLED`?), so you could verify nothing has been created in that location.

Reply:
I added in AbstractTestHive, please review, thanks.